server/src/main/java/org/elasticsearch/index/mapper/DocumentMapper.java

```diff
@@ -72,9 +88,28 @@ public static DocumentMapper createEmpty(MapperService mapperService) {
             : "provided source [" + source + "] differs from mapping [" + mapping.toCompressedXContent() + "]";
     }
 
-    private void maybeLogDebug(Exception ex) {
+    private void maybeLog(Exception ex) {
         if (logger.isDebugEnabled()) {
             logger.debug("Error while parsing document: " + ex.getMessage(), ex);
+        } else {
+            final long now = Instant.now().getEpochSecond();
+            // Check without locking first, to reduce lock contention.
+            if (now - lastErrorLogEpochSecond > errorThrottlingIntervalSeconds) {
+                boolean shouldLog = false;
+                // Acquire spinlock.
+                while (lastErrorLogLock.compareAndSet(false, true)) {
+                }
+                // Repeat check under lock, so that only one message gets written per interval.
+                if (now - lastErrorLogEpochSecond > errorThrottlingIntervalSeconds) {
+                    shouldLog = true;
+                    lastErrorLogEpochSecond = now;
+                }
+                // Release spinlock before logging.
+                lastErrorLogLock.set(false);
+                if (shouldLog) {
+                    logger.error("Error while parsing document: " + ex.getMessage(), ex);
+                }
```

Comment:
Good point, a spinlock is not needed here since we only expect one op to succeed. The one that succeeds will log, any failing `compareAndSet` calls would lead to the condition failing right after. Done.